man/checkdirfile.1
.\" checkdirfile.1.  The checkdirfile man page.
.\"
.\" Copyright (C) 2010, 2012 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH checkdirfile 1 "21 February 2012" "Version 0.8.0" "GETDATA"
.SH NAME
checkdirfile \(em verify Dirfile database metadata
.SH SYNOPSIS
.nh
.ad l
.P
.B checkdirfile
.I DIRFILE
.P
.B checkdirfile
.RB "[\| " \-\-help " | " \-\-version " \|]"
.hy
.ad n
.SH DESCRIPTION
Report errors in metadata of the dirfile(5) database specified by 
.IR DIRFILE .

Three separate checks are performed.  First, the format specification metadata
is scanned for syntax errors (see dirfile-format(5)) and Dirfile Standards
compliance.  Second, fields defined in the database are checked for
self-consistency.  Finally, the database is queried for its length.
.P
.B --help
.PD 0
.TP
.PD
.B --version
Show usage and version information and exit.

.SH AUTHOR

checkdirfile was written by D. V. Wiebe.

.SH REPORTING BUGS

Please send reports of bugs to getdata\-devel@lists.sourceforge.net

The GetData home page: <http://getdata.sourceforge.net/>

.SH COPYRIGHT

Copyright \(co 2010 D. V. Wiebe

checkdirfile is licenced under the GNU LPGL version 2.1 or later.

This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
.BR dirfile (5),
.BR dirfile-format (5)


NEWS
New in version 0.8.0a:

  * Some stuff.

|==============================================================================|

New in version 0.7.3:

  Library Changes

  * BUG FIX: Passing a zero dividend to alter_crecip() and alter_recip() now
    results in the dividend not changing, per documentation.
  * BUG FIX: GetData wasn't properly computing the module directory, causing
    encodings using external modules to fail with GD_E_UNSUPPORTED.
  * BUG FIX: Metafield records are now properly stored in the format file for
    Standards Version 8.
  * BUG FIX: gd_raw_filename() wasn't clearing the Dirfile error before
    operation, resulting in it failing erroneously in certain situations.

  Utilities Changes:

  * BUG FIX: A potential segmentation fault has been patched in dirfile2ascii.

|==============================================================================|

New in version 0.7.2:

  Library Changes

  * If built with modules, the plugin shared objects will now be installed in
    a separate directory (by default "${libdir}/getdata", but see the
    --with-module-dir option in ./configure).
  * BUG FIX: When using modules, a missing plugin no longer results in a lock up
    the second time an attempt is made to find it.
  * BUG FIX: GetData now properly processes /INCLUDE directives with absolute
    paths, as the documentation insists it should.
  * BUG FIX: gd_add()ing the first raw field in a dirfile to a subfragment no
    longer results in memory corruption when calling gd_close().

  API Changes

  * If called with GD_CREAT but no encoding specified, gd_include() now will
    duplicate the encoding of the parent fragment (if any).

  Miscellaneous

  * The python module install dir can now be changed by passing
    --with-python-module-dir to configure.

|==============================================================================|

New in version 0.7.1:

  Library Changes

  * BUG FIX: When the parser fails in a gd_include() call, the internal list of
    conformant Standards Versions for the loaded dirfile is no longer corrupted.
  * BUG FIX Performing two consecutive reads of the same field in an ASCII
    encoded datasources no longer confuses GetData.
  * BUG FIX: Numerous memory leaks have been patched (mostly pin-holes, although
    there are a few, rare but egregious ones).

  API Changes:

  * The value(s) of a CONST or CARRAY field added with gd_[m]add() is now
    guaranteed to be zero; previously, it was unspecified.
  * BUG FIX: Specifying both GD_VERBOSE and GD_IGNORE_DUPS no longer results in
    spurious "Field code already defined" messages when duplicate fields are
    encountered.
  * BUG FIX: Calling gd_delete on a metafield no longer results in a segfault.

  Bindings Changes:

  * F77 BUG FIX: A memory leak in GDASCA has been fixed.
  * Python BUG FIX: The first element of a python list (instead of a NumPy
    array) is no longer dropped when passed to pygetdata on 64-bit systems.

  Utilities Changes:

  * dirfile2ascii now pads reads past the end-of-field.  The value used to
    pad may be specified with '-z', and defaults to 0/NaN.  Previously, the
    value of uninitialized memory was output.
  * BUG FIX: dirfile2ascii now does what the user intends when presented with
    cart-before-horse arguments, ie. "dirfile2ascii -e field DIRFILE".
    Previously this resulted in confusion.
  * BUG FIX: dirfile2ascii now accepts field conversion '-u', which was
    documented but overlooked in the argument parser.

|==============================================================================|

New in version 0.7.0:

  Dirfile Changes

  * Dirfile Standards Version 8 has been released.  It adds three new field
    types: DIVIDE, which computes X/Y for vectors X and Y, RECIP, which computes
    A/X for vector X and scalar A, and CARRAY, which is an array of CONST
    scalars.  A few other changes are listed below.
  * Like CONST fields, CARRAY elements can be used as parametersin the
    definition of other fields.  Syntax is 'field<n>', where 'field' is the
    field name and 'n' the element index in the CARRAY (and the angle brackets
    are literal angle brackets).  If the '<n>' is omitted, the first element
    (numbered 0) is assumed.
  * Slashes on reserved words are now mandatory, allowing fields with reserved
    words as names.
  * The long-deprecated single character data type codes have been removed from
    Standards Version 8.  (They are still accepted by GetData, which handles
    all previous Standards Versions as well.)
  * The /ENDIAN directive now takes an optional second parameter.  The only
    allowed value for this second token is "arm", which indicates that double
    precision floating point data (including double precision complex data) are
    stored in the ARM middle-endian format.  Without this, floating point data
    are assumed to have the same byte sex as integer data.

  Library Changes

  * GetData, including its bindings, is now supported under MacOS X, Cygwin, and
    Win32 using MinGW.  In the case of MinGW, this includes linking against the
    Microsoft C Runtime, which isn't POSIX compliant.  Users should excercise
    caution when writing to dirfiles in this case.
  * GetData can now be used on ARM processors.  As a side-effect, GetData can
    now read and write the middle-ended floating point format used by older ARM
    processors (on any processor, via endianness conversion).
  * Some changes have been made to the parser.  It now operates in one of two
    modes: pedantic mode, which adheres strictly to one particular Dirfile
    Standards Version, and permissive mode (the default) which is more
    lenient, as the old parser was.  Encountering a /VERSION directive will
    switch the parser from permissive to pedantic mode unless told not to (see
    GD_PERMISSIVE below), and will change the particular Standards Version to
    which pedantic mode is adhering.

util/checkdirfile.c
/* Copyright (C) 2007-2010 D. V. Wiebe
 *
 ***************************************************************************
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "getdata.h"

static int callback(gd_parser_data_t *pdata, void* ne)
{
  char getdata_error[2048];
  printf("  syntax error: %s\n", gd_error_string(pdata->dirfile,
        getdata_error, 2048));

  (*(int*)ne)++;

  return GD_SYNTAX_IGNORE;
}

int main(int argc, char* argv[])
{
  DIRFILE* dirfile;
  int e, i, j;
  char getdata_error[2048];
  off_t n;
  int ne = 0;
  unsigned int nfields;
  const char **flist, **mflist;

  if (argc < 2 || !strcmp(argv[1], "--version") || !strcmp(argv[1], "--help")) {
    printf("Usage:\n"
        "  checkdirfile DIRFILE                Check the DirFile database "
        "DIRFILE for\n"
        "                                        errors.\n"
        "  checkdirfile [ --help | --version]  Print this message and exit.\n"
        "\n\n"
        "This program is part of %s.  Copyright (C) 2008-2010  D. V. Wiebe\n"
        "Please send reports of bugs and other communication to:\n  %s\n\n"
        "This program comes with NO WARRANTY, not even for MERCHANTABILITY "
        "or FITNESS\n"
        "FOR A PARTICULAR PURPOSE. You may redistribute it under the terms of "
        "the GNU\n"
        "Lesser General Public License, either version 2.1 of the License, or "
        "(at your\n"
        "option) any later version.\n\n"
        "You should have received a copy of the GNU Lesser General Public "
        "License along\n"
        "with this program; if not, write to the Free Software Foundation,"
        "Inc.,\n"
        "51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA\n",
        PACKAGE_STRING, PACKAGE_BUGREPORT);
    return 1;
  }

  /* open the dirfile */
  puts("Checking syntax...");
  dirfile = gd_cbopen(argv[1], GD_RDONLY, callback, &ne);

  /* open error */
  e = gd_error(dirfile);
  if (e) {
    if (e != GD_E_FORMAT) {
      printf("  getdata error: %s\n", gd_error_string(dirfile, getdata_error,
            2048));
      gd_close(dirfile);
      return 1;
    }
  }

  if (ne > 0)
    printf("  Found %i line%s with syntax errors.\n", ne, (ne == 1) ? "" : "s");
  else {
    int vers[GD_DIRFILE_STANDARDS_VERSION + 1];
    int nvers = 0;
    int first = -1;
    int start = 1;

    printf("  Syntax OK.\n\n");

    for (i = 0; i <= GD_DIRFILE_STANDARDS_VERSION; ++i) {
      if (gd_dirfile_standards(dirfile, i) == i) {
        vers[i] = 1;
        nvers++;
      } else
        vers[i] = 0;
    }

    if (nvers == 0) {
      puts("WARNING: Dirfile conforms to no Standards Version.");
    } else {
      printf("Dirfile conforms to Standards %s ",
          (nvers == 1) ? "Version" : "Versions");

      for (i = 0; i <= GD_DIRFILE_STANDARDS_VERSION; ++i) {
        if (vers[i]) {
          if (first == -1)
            first = i;
        } else if (first != -1) {
          if (!start)
            fputs(", ", stdout);
          else
            start = 0;

          if (first == i)
            printf("%i", i);
          else if (first + 1 == i)
            printf("%i, %i", first, i);
          else
            printf("%i-%i", first, i);

          first = -1;
        }
      }

      if (first != -1) {
        if (!start)
          fputs(", ", stdout);

        if (first == GD_DIRFILE_STANDARDS_VERSION)
          printf("%i", GD_DIRFILE_STANDARDS_VERSION);
        else if (first + 1 == GD_DIRFILE_STANDARDS_VERSION)
          printf("%i, %i", first, GD_DIRFILE_STANDARDS_VERSION);
        else
          printf("%i-%i", first, GD_DIRFILE_STANDARDS_VERSION);

        fputs(" (the latest version)", stdout);
      }

      puts("");
    }

  }

  /* Check the validity of each field defined */
  ne = 0;
  puts("\nChecking fields...");
  flist = gd_field_list(dirfile);
  for (i = 0; flist[i] != NULL; ++i) {
    if (gd_validate(dirfile, flist[i])) {
      printf("  getdata error checking %s: %s\n", flist[i],
          gd_error_string(dirfile, getdata_error, 2048));
      ne++;
    }
    mflist = gd_mfield_list(dirfile, flist[i]);
    for (j = 0; mflist[j] != NULL; ++j) {
      char code[GD_MAX_LINE_LENGTH];
      snprintf(code, GD_MAX_LINE_LENGTH, "%s/%s", flist[i], mflist[j]);
      if (gd_validate(dirfile, code)) {
        printf("  getdata error checking %s: %s\n", code,
            gd_error_string(dirfile, getdata_error, 2048));
        ne++;
      }
    }
  }

  nfields = gd_nfields(dirfile);

  if (ne > 0)
    printf("  Found %i problems in %u %s.\n", ne, nfields,
        nfields > 1 ? "fields" : "field");
  else
    printf("  No problems found in %u %s.\n", nfields,
        nfields > 1 ? "fields" : "field");

  /* try to retrieve the number of frames in the dirfile */
  puts("\nChecking frames...");
  n = gd_nframes(dirfile);

  if (gd_error(dirfile)) {
    printf("  getdata error: %s\n", gd_error_string(dirfile, getdata_error,
          2048));
    gd_close(dirfile);
    return 1;
  }

  printf("  Found %llu frames.\n", (unsigned long long)n);

  gd_close(dirfile);
  return 0;
}
